### DISCUSSION TITLE
I cannot find any documentation or examples or tests showing how Nats-Rollup is intended to work

### DISCUSSION THREAD

**Discussion by tpihl** [nats-io/nats-server#4417] [Q&A] [answered]
I know there has been information about all headers that jetstream considers serverside,

Same thing for the Nats-Expected-*

Where did that go, and how should they be used. We need both rollup and Expected seqno per subject

/T

**Comment by derekcollison** [comment]
@bruth can direct towards docs.

In server, we have the following..

<img width="558" alt="image" src="https://github.com/nats-io/nats-server/assets/90097/d05b6f87-37bb-4058-96c6-4485c9b95ad0">

**Reply by tpihl** [reply]
@derekcollison  I saw the code and I was looking for tests that tested this in server (to see how to use it), but i could not find any. Maybe there should be?

**Reply by ripienaar** [reply]
TestJetStreamRollup, TestJetStreamClusterRollupSubjectAndWatchers, TestJetStreamClusterRollups, TestJetStreamClusterRollupsRequirePurge

**Comment by bruth** [comment]
@tpihl Good call-out. Here is a [PR](https://github.com/nats-io/nats.docs/pull/610/files) for rollup to the current docs. Will add another for `Nats-Expected-*`. The short answer is that the header will result in the server enforcing optimistic concurrency control at the stream or subject level (within the stream). For example, if two concurrent processes attempt to publish to a stream with subject `foo.1` (if scoped to the subject) and both specify an expected sequence of `2`, the one received first will be accepted and the other rejected.

**Reply by tpihl** [reply]
@bruth Looks great for doc. Maybe also som small table somewhere with all headers that the stream will consider at the same place (dedup, optimistic lock, rollup and, yeah, the whole list Derek show above.

**Reply by bruth** [reply, marked as answer]
👍 Indeed, that is coming next as well.
